woocommerce/woocommerce
# Fix filter by attribute widget now working for "Any..." attributes

### All Submissions:


### Changes proposed in this Pull Request:

This pull requests fixes two bugs related to attribute filtering for variable products:

1. [Filter by attribute widget not counting variations having an attribute value of "Any..."](https://github.com/woocommerce/woocommerce/issues/27419)

When a variation has a value of "Any..." for a given attribute, then that variation isn't counted by the "Filter by attribute" widget corresponding to that attribute.

2. `attribute_` metadata entries with empty values not handled correctly when adding or removing variation attributes

This is an old bug (present in pre-4.4 versions). When a new variation attribute is added to the main product, then all the already existing variations will display the attribute with a value of "Any...", but the corresponding `attribute_` metadata entry for the variation won't be created. (If an existing variation is manually set to "Any..." and saved then it works as expected and the entry is created)

Conversely, when removing the attribute from the main product the corresponding `attribute_` metadata entries for the variations won't be removed.

A data migration is added too to backfill possible missing values for these metadata entries.

Closes #27419.

### How to test the changes in this Pull Request:

Start with an empty shop (no products).

1. Create an attribute named "Color" with three terms, e.g. "Black", "White", "Blue".
2. Create an attribute named "Style" with three terms, e.g. "Classic", "Sport", "Beach".
3. Add a "Filter by attribute" widget, to filter by "Color".
4. Add a variable product in which the "Color" attribute is used for all variations, with one variation per color.
5. Load the shop, verify that the filter by color widget is displaying all three colors, with a count of 1 for each.
6. Add a second variable product, but this time use the "Style" attribute to generate all variations.
7. After the three style variations have been created, add "Color" as an additional variation attribute with all three values. Now the existing variations should automatically get "Any..." as the value for the color attribute:

![image](https://user-images.githubusercontent.com/937723/91537309-5c945a00-e916-11ea-8fe3-792cd632eb49.png)

8. Reload the shop. Without the fix, it still displays 1 for all colors. With the fix, it displays 2 for each color (it's counting the "Any..." product now).
9. Set one of the variations to "Out of stock". Reload the shop and verify that the value of the counters hasn't changed.
10. Set all the variations to "Out of stock". Reload the shop and verify that the value of the counters is now 1 (and now the product doesn't display in the shop). Set the variations to "In stock" again.
11. Remove the "Color" attribute from the product, reload the shop and verify that all counters go back to 1.

### Other information:


### Changelog entry

> Fix - Filter by attribute widget not working properly for variations having attribute values of "Any..."
> Fix - "attribute_" metadata entries not created and deleted properly for variations
